src/space.cpp

```diff
@@ -435,9 +340,128 @@ void makeNaCl(Space &space, int num_particles, const Geometry::Chameleon &geomet
 
     json j = json::array();
     j.push_back({{"salt", {{"N", num_particles}}}});
-    Faunus::insertMolecules(j, space);
+    InsertMoleculesInSpace::insertMolecules(j, space);
 }
 
 } // namespace SpaceFactory
 
+bool InsertMoleculesInSpace::insertAtomicGroups(MoleculeData &moldata, Space &spc, int N, bool inactive) {
+    assert(moldata.atomic == true);
+    typename Space::Tpvec p;
+    p.reserve(N * moldata.atoms.size());
+    while (N-- > 0) {
+        auto particles = moldata.getRandomConformation(spc.geo, spc.p);
+        p.insert(p.end(), particles.begin(), particles.end());
+    }
+    spc.push_back(moldata.id(), p);
+    if (inactive) {
+        spc.groups.back().resize(0);
+    }
+    return true;
+}
+void InsertMoleculesInSpace::insertMolecularGroups(MoleculeData &moldata, Space &spc, int num_molecules,
+                                                   bool inactive) {
+    assert(moldata.atomic == false);
+    while (num_molecules-- > 0) { // insert molecules
+        spc.push_back(moldata.id(), moldata.getRandomConformation(spc.geo, spc.p));
+        if (inactive) {
+            spc.groups.back().unwrap(spc.geo.getDistanceFunc());
+            spc.groups.back().resize(0);
+        }
+    }
+}
+bool InsertMoleculesInSpace::setPositionsForTrailingGroups(Space &spc, int num_molecules,
+                                                           const Faunus::ParticleVector &particles,
+                                                           const Point &offset) {
+    assert(spc.groups.size() >= num_molecules);
+    if (particles.size() == num_molecules * (spc.groups.end() - num_molecules)->traits().atoms.size()) {
+        size_t cnt = spc.p.size() - particles.size();
```

Comment:
A more descriptive variable name please. It is more like a particle index than count. After changing the loop it will be a offset of the particle vector.